docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,1882 @@
+| Go 2 (presumably new types)                    | BAD       | BAD         | ?                      | BAD     | ?            | BAD                  |
+| (CRDB) `roachpb.Error`                         | BAD       | good        | BAD                    | BAD     | BAD          | BAD                  |
+| (CRDB) `distsqlpb.Error`                       | BAD       | good        | BAD                    | good    | BAD          | BAD                  |
+| (CRDB) `pgerror.Error` (2.1/previous)          | BAD       | good        | good                   | good    | good         | good                 |
+| (CRDB) proposed new error objects              | good      | good        | good                   | good    | good         | good                 |
+
+This failure by the current code to meet all our requirements is the main motivation for this work.
+
+# Guide-level explanation
+
+The package is `github.com/cockroachdb/cockroach/pkg/errors`.
+
+Table of contents:
+
+- [Instantiating new errors](#Instantiating-new-errors)
+- [Decorating existing errors](#Decorating-existing-errors)
+- [Utility features](#Utility-features)
+  - [Adding `context`](#Adding-context)
+  - [Formatting variants](#Formatting-variants)
+  - [Safe details](#Safe-details)
+  - [Additional PostgreSQL error fields](#Additional-PostgreSQL-error-fields)
+  - [Telemetry keys](#Telemetry-keys)
+  - [Depth variants](#Depth-variants)
+- [Handling chains of error causes](#Handling-chains-of-error-causes)
+  - [Accessing the cause](#Accessing-the-cause)
+  - [Preservation of causes across the wire](#Preservation-of-causes-across-the-wire)
+  - [Barrier errors](#Barrier-errors)
+  - [Multiple causes](#Multiple-causes)
+  - [Identification of causes](#Identification-of-causes)
+  - [Error equivalence and markers](#Error-equivalence-and-markers)
+- [Promotion and preservation during formatting](#Promotion-and-preservation-during-formatting)
+- [What comes out of an error?](#What-comes-out-of-an-error)
+
+## Instantiating new errors
+
+Instantiating a new error can be as simple as `errors.New("hello")` or
+`errors.Errorf("hello %s", "world")`.
+
+To equip a code useful to PostgreSQL clients when constructing a new
+error, the following form is preferred:
+`errors.New("hello").WithDefaultCode(pgerror.CodeSyntaxError)`
+
+The library is compatible with existing protobuf error objects, so
+instantiating, for example, with `err :=
+&roachpb.RangeRetryError{Reason: "hello"}` is also valid.
+
+## Decorating existing errors
+
+Adding some words of context can be as simple as `errors.Wrap(err,
+"hello")`, although the form `Wrapx()` is preferred (see next section).
+
+To add a code useful to PostgreSQL clients, one can use
+e.g. `errors.WithDefaultCode(err, pgerror.CodeSyntaxError)`.
+The new code is only used if the original error did not provide a code already.
```

Comment:
Are these the right semantics? I suppose morally you want to not wrap an error with an error code twice. How about a boolean return or chaining with a marker error (so that you keep both codes, you can then decide if you want the inner our outer)?